### PR TITLE
Initial contribution of the Exec 2.0 binding.

### PR DESCRIPTION
**Pull request by kgoderis**
Signed-off-by: Karel Goderis karel.goderis@me.com

### CODE REVIEW

**Comment by stedaniels** [comment]
Is anyone using this in "production"?
I'll try and find some time to test it out.  Migrating to OH2 soon!

**Comment by kgoderis** [comment]
@stedaniels I am

**Comment by ThomDietrich** [comment]
no README.md? Would be great to see this addition merged.

File: addons/binding/org.openhab.binding.exec/ESH-INF/binding/binding.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="exec"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:binding="http://eclipse.org/smarthome/schemas/binding/v1.0.0"
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+    <name>Exec Binding</name>
+    <description>This is the binding for Exec.</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/ESH-INF/binding/binding.xml]: please improve the desc like "... for executing commands on the command line."

File: addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="exec"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+    <thing-type id="command">
+        <label>Command</label>
+        <description>The Command Thing encapsulates a shell command to be executed</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml]: remove the word "Thing" here

File: addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        </channels>
+        
+        <config-description>
+            <parameter name="command" type="text" required="true">
+                <label>Command</label>
+                <description>The command to execute</description>
+            </parameter>
+            <parameter name="transform" type="text" required="false">
+                <label>Transform</label>
+                <description>The transformation to apply on the execution result</description>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml]: maybe provide an example string here?

File: addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        <description>The exit value of the last execution of the command</description>
+        <state readOnly="true"></state>  
+    </channel-type>
+     <channel-type id="running">
+        <item-type>Switch</item-type>
+        <label>Running</label>
+        <description>Indicator whether the command is currently being executed</description>
+        <state readOnly="true"></state>  
+    </channel-type>
+     <channel-type id="execute">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml]: I'd remove this channel and combine it with "running" into a "run" channel. You can send ON to execute the command and the current state tells you whether it is running or not.

File: addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        <item-type>Switch</item-type>
+        <label>Execute</label>
+        <description>When triggered with by ON, the command will be executed</description>
+    </channel-type>
+    <channel-type id="lastexecution">
+        <item-type>DateTime</item-type>
+        <label>Last Execution</label>
+        <description>Time/Date the command was last executed</description>
+        <state readOnly="true"></state>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/ESH-INF/thing/thing-types.xml]: provide a pattern here

File: addons/binding/org.openhab.binding.exec/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,26 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Exec Binding
+Bundle-SymbolicName: org.openhab.binding.exec;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ org.apache.commons.exec;version="1.1.0",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/META-INF/MANIFEST.MF]: remove version constraints

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    public static final String INTERVAL = "interval";
+    public static final String TIME_OUT = "timeout";
+    public static final String COMMAND = "command";
+    public static final String TRANSFORM = "transform";
+
+    /** RegEx to extract a parse a function String <code>'(.*?)\((.*)\)'</code> */
+    private static final Pattern EXTRACT_FUNCTION_PATTERN = Pattern.compile("(.*?)\\((.*)\\)");
+
+    private ScheduledFuture<?> executionJob;
+    // ByteArrayOutputStream stdout;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: remove

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            String commandLine = (String) getConfig().get(COMMAND);
+
+            int timeOut = 60000;
+            if (((BigDecimal) getConfig().get(TIME_OUT)) != null) {
+                timeOut = ((BigDecimal) getConfig().get(TIME_OUT)).intValue() * 1000;
+            }
+
+            if (commandLine != null && !commandLine.isEmpty()) {
+
+                updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.RUNNING), OnOffType.ON);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: replace by `updateState(ExecBindingConstants.RUNNING, OnOffType.ON);`

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                // problem for external commands that generate a lot of output, but this will be dependent on the limits
+                // of the underlying operating system.
+
+                Process proc = null;
+                try {
+                    proc = rt.exec(commandLine.toString());
+                } catch (Exception e) {
+                    logger.error("An exception occured while executing '{}' : '{}'",
+                            new Object[] { commandLine.toString(), e.getMessage() });
+                    updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.RUNNING), OnOffType.OFF);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                    updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.OUTPUT),
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+                try {
+                    InputStreamReader isr = new InputStreamReader(proc.getInputStream());
+                    BufferedReader br = new BufferedReader(isr);
+                    String line = null;
+                    while ((line = br.readLine()) != null) {
+                        outputBuilder.append(line).append("\n");
+                        logger.debug("Exec [{}]: '{}'", "OUTPUT", line);
+                    }
+                    br.close();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: you should do the closing in the finally block or use try-with-resources instead.

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+
+                try {
+                    InputStreamReader isr = new InputStreamReader(proc.getErrorStream());
+                    BufferedReader br = new BufferedReader(isr);
+                    String line = null;
+                    while ((line = br.readLine()) != null) {
+                        errorBuilder.append(line).append("\n");
+                        logger.debug("Exec [{}]: '{}'", "ERROR", line);
+                    }
+                    br.close();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+                } catch (IOException e) {
+                    logger.error("An exception occured while reading the stderr when executing '{}' : '{}'",
+                            new Object[] { commandLine.toString(), e.getMessage() });
+                }
+
+                boolean exitVal = false;
+                try {
+                    exitVal = proc.waitFor(timeOut, TimeUnit.MILLISECONDS);
+                } catch (InterruptedException e) {
+                    logger.error("An exception occured while waiting for the pocess ('{}') to finish : '{}'",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: typo: pocess

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+                            new Object[] { commandLine.toString(), e.getMessage() });
+                }
+
+                if (!exitVal) {
+                    logger.warn("Forcibly termininating the pocess ('{}') after a timeout of {} ms",
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: typo: pocess

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+                            new Object[] { commandLine.toString(), timeOut });
+                    proc.destroyForcibly();
+                }
+
+                updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.RUNNING), OnOffType.OFF);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+                updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.EXIT),
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+                        new DecimalType(proc.exitValue()));
+
+                String transformedResponse = StringUtils.chomp(outputBuilder.toString());
+                String transformation = (String) getConfig().get(TRANSFORM);
+
+                if (transformation != null && transformation.length() > 0) {
+                    transformedResponse = transformResponse(transformedResponse, transformation);
+                }
+
+                updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.OUTPUT),
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

File: addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+                        new StringType(transformedResponse));
+
+                DateTimeType stampType = new DateTimeType(Calendar.getInstance());
+                updateState(new ChannelUID(getThing().getUID(), ExecBindingConstants.LAST_EXECUTION), stampType);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java]: see above

**Comment by kgoderis** [comment]
@kaikreuzer The changes are there, I am only struggling to squash the pom.xml update into the commit. It is commit-hell in SourceTree at my side, so if you do not mind, maybe you can integrate the 2 commits?

**Comment by kaikreuzer** [comment]
Thanks, looks all good!

I cannot see a second commit from you though. The bindings/pom.xml update is still missing, likewise the addition in the feature.xml. Feel free to push that as a second commit to your branch, no need to squash - I can do so while merging.

**Comment by kgoderis** [comment]
> Thanks, looks all good!
> 
> I cannot see a second commit from you though. The bindings/pom.xml update is still missing, likewise the addition in the feature.xml. Feel free to push that as a second commit to your branch, no need to squash - I can do so while merging.

I had it but removed it afterwards because jenkins was spamming the hell out of me with compile problems. I will create a new clean branch and do a PR from there.

**Comment by kaikreuzer** [comment]
No need, I can also merge this one and you just create one with the pom and feature.

**Comment by kgoderis** [comment]
works for me

**Comment by kaikreuzer** [comment]
Looking forward to your next PR then ;-)

**Comment by monnimeter** [comment]
Is there a way to force OH2 to install the 1.x binding if addressed via addons.cfg?

With
````
binding = exec, ...
````
the newly merged version is installed after ``apt-get upgrade``, even if the legacy support is enabled. This makes some of my items unusable...

I would switch to this version, if there's a way to pass options to the script like here
```
send_cmd.sh 192.168.178.22 setBetriebsart_A1 %2$s
````

Thanks and regards,
Osman

**Comment by kaikreuzer** [comment]
Distro 615 correctly installs the 1.x binding if you use "exec1" in your list.

**Comment by kaikreuzer** [comment]
Oops, wrong - it doesn't. Build 616 will do it, which includes https://openhab.ci.cloudbees.com/job/openHAB1-Addons/1345/changes

**Comment by kgoderis** [comment]
> I would switch to this version, if there's a way to pass options to the script like here

@monnimeter This should work. In fact, the complete string is passed on to the shell

**Comment by monnimeter** [comment]
@kaikreuzer it works perfectly. Thank you for the hint.

@kgoderis you're right, the string is passed as it is... But I would need a concept to pass the state of an item to the shell :)

I'm using an item to switch the operation mode ("Betriebsart") of my heating ("Heizung") with this item defined:
```
Number  HZ_getBetriebsart_A1  "Betriebsart Heizung [MAP(E_BA2.map):%s]"  <ba>  (HZ_A1)  { exec = ">[*:/opt/viessmann/send_cmd.sh 127.0.0.1 setBetriebsart_A1 %2$s]" }
```

The pattern ``%2$s`` is substituted by the current state (or command) of the item (java.util.Formatter syntax is applied). Within the sitemap I only allow the states 0 and 3...
```
Switch item=HZ_getBetriebsart_A1 mappings=[0="Nur WW", 3="H+WW"]
```
... hence the shell script is executed like ``setBetriebsart_A1 0`` or ``setBetriebsart_A1 3`` depending on which button was pressed.

A possible solution for those use cases could be the introduction of 
* a string token which can be used within the command parameter of the thing configuration like
```Thing exec:command:xyz [ command="/usr/local/bin/xyz  %2$s", interval=15, timeout=5]```
* an additional channel ``input`` of type ``String``
```String XYZParam "[%s]" (All) { channel="exec:command:xyz:input" }```

During each execution, the token should be substituted by the actual state of the input. Due to the fact that the exec2 binding supports repeated execution, this would make *the thing* really powerfull :)

What do you think about this?

Thank you much and regards,
Osman

**Comment by kgoderis** [comment]
@monnimeter That is what I thought you mentioned in the initial post ;-)

The way it is currently set up (e.g. execution is triggered by ON on a Switch compatible Channel), we should go for suggestion 2. That being said, how many parameters do we need?

The input string could be comma separated list of values that are based on to argument %1, %2, ... and so forth.

**Comment by monnimeter** [comment]
I hope I'm getting the question right :)

IMHO we should stick with the concept of the exec1 binding. It supports 3 argument types which can be passed to the script in any order (even multiple times):
* %1 -> the current date (as java.util.Date, example: %1$tY-%1$tm-%1$td)
* %2 -> the current command or state (example: %2$s)
* %3 -> the current item name (example: %3$s)

So a command thing could be configured with
```
Thing exec:command:cmd [ command="/opt/cmd.sh %2$s %1$tY-%1$tm-%1$td %3$s", interval=15, timeout=5 ]
```

The input channel of that thing could be linked to an item, so %2 and %3 can be derived. %1 is set to ``now()`` when the channel ``run`` is triggered. Example:
```
Switch item1 "my item [%s]" (All) { channel="exec:command:xyz:input" }
```
The script would be then called with i.e. ``/opt/cmd.sh ON 2016-11-27 item1``


I'm not really sure, but IMHO we could also support multiple links to the channel like
```
Switch item1 "my 1st. item [%s]" (All) { channel="exec:command:xyz:input" }
Switch item2 "my 2nd. item [%s]" (All) { channel="exec:command:xyz:input" }
```
When triggered, the script should be called multiple times with i.e. 
```
/opt/cmd.sh ON 2016-11-27 item1
/opt/cmd.sh ON 2016-11-27 item2
```
I don't know if there's a way to iterate through the linked items of a channel and if this concept is compliant to the ESH philosophy. But maybe you know that and / or @kaikreuzer could say a word or two regarding these suggestions.

One more idea: we could also introduce a new thing configuration parameter``autoru``. If set to ``true`` the script is executed on every change of the corresponding linked input item. The already existing "run" channel would then show if the script is running or not.

**Comment by kgoderis** [comment]
mmh.... You rightly identified the issue with respect to linked items. In essence, the Binding does not know (and should not) about linked Items at all. As far as it is concerned, it is only knowing about Channels (that, in turn could have a State / value that can be leveraged here). And beyond that, a Channel can be linked to several items, so it would imply that the shell command is executed multiple times

Will add the autorun idea.

**Comment by monnimeter** [comment]
Ooohkay... I see. Then let's think about the simple best can do approach which IMHO is having just 2 argument types:
* %1 -> the current date (as java.util.Date, example: %1$tY-%1$tm-%1$td)
* %2 -> the current command or state of the input channel (example: %2$s)

Not having the above mentioned argument type %3 makes linking multiple items to the input channel almost senseless... as we cannot distinguish which channel was the trigger. 

What do you thin... does this make more sense?

BTW: thank you for implementing the autorun feature... it makes some rules obsolete.

**Pull request merged**